### PR TITLE
[W-10259197/W-10259190] Automate Account Naming Test Suite

### PR DESCRIPTION
**Pull request by ZeeshanSyedH**
This pull requests automates the account naming logic as it relates to administrative and household accounts.

# Issues Closed
- [W-10259190](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000UngHlYAJ/view)
- [W-10259197](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000UngfxYAB/view)

# Testing Notes
- [T-3904053](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZdVIAU/view)
- [T-3904055](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZdfIAE/view)
- [T-3904056](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZdkIAE/view)
- [T-3904057](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZdpIAE/view)
- [T-3904058](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZduIAE/view)
- [T-3904059](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZdlIAE/view)
- [T-3904061](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001Ze4IAE/view)
- [T-3904063](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeEIAU/view)
- [T-3904064](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeJIAU/view)
- [T-3904065](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeOIAU/view)
- [T-3904066](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeTIAU/view)
- [T-3904067](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeYIAU/view)
- [T-3904068](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZedIAE/view)
- [T-3904070](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZenIAE/view)
- [T-3904075](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZfCIAU/view)
- [T-3904076](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZfHIAU/view)

### CODE REVIEW

**Review by iskander-sfdc — CHANGES_REQUESTED** [review]
Functional test classes work in orgs where EDA managed package is installed. We need to add the namespace to the below classes:
```Error: Could not process MDAPI response: Update of ApexClass ACCT_NameProcessing_TEST: Error on line 45, col 13: Invalid type: Hierarchy_Settings__c

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 52, col 13: Variable does not exist: UTIL_CustomSettingsFacade

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 54, col 27: Variable does not exist: UTIL_UnitTestData_API

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 72, col 22: Variable does not exist: ACCT_AdministrativeNameRefresh_TEST

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 44, col 22: Variable does not exist: ACCT_IndividualAccounts_TEST

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 80, col 13: Variable does not exist: UTIL_CustomSettingsFacade

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 82, col 27: Variable does not exist: UTIL_UnitTestData_API

Update of ApexClass ACCT_NameProcessing_TEST: Error on line 73, col 13: Invalid type: Hierarchy_Settings__c
```

**Comment by ZeeshanSyedH** [comment]
> Functional test classes work in orgs where EDA managed package is installed. We need to add the namespace to the below classes:
> 
> ```
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 52, col 13: Variable does not exist: UTIL_CustomSettingsFacade
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 54, col 27: Variable does not exist: UTIL_UnitTestData_API
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 72, col 22: Variable does not exist: ACCT_AdministrativeNameRefresh_TEST
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 44, col 22: Variable does not exist: ACCT_IndividualAccounts_TEST
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 80, col 13: Variable does not exist: UTIL_CustomSettingsFacade
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 82, col 27: Variable does not exist: UTIL_UnitTestData_API
> 
> Update of ApexClass ACCT_NameProcessing_TEST: Error on line 73, col 13: Invalid type: Hierarchy_Settings__c
> ```

Have alternatively redistributed the unpackaged tests into the managed package ([reference commit](https://github.com/SalesforceFoundation/EDA/pull/1950/commits/a35d655bfd5049392bc56ad4e30aeab3314c959b)).

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls [hunk 1]
```diff
@@ -191,6 +191,12 @@ public class ACCT_IndividualAccounts_TDTM extends TDTM_Runnable {
             dynamicSoql += 'CurrencyIsoCode, ';
         }
 
+        Set<String> objectFields = Schema.SObjectType.Contact.fields.getMap().keySet();
+        Boolean ispersonNameSuffixEnabled = objectFields.contains('suffix');
+        if (ispersonNameSuffixEnabled) {
+            dynamicSoql += 'Suffix, ';
+        }
+
```
Comment (iskander-sfdc) [inline force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:199]: Could you please clarify what is the purpose of this change to TDTM? I thought this is just a test automation story

File: force-app/main/default/classes/UTIL_ACCT_Naming.cls [hunk 2]
```diff
@@ -165,6 +165,12 @@ public class UTIL_ACCT_Naming {
             'OtherPostalCode, OtherCountry, OtherLatitude, OtherLongitude, ' +
             'Phone, Fax';
 
+        Set<String> objectFields = Schema.SObjectType.Contact.fields.getMap().keySet();
+        Boolean ispersonNameSuffixEnabled = objectFields.contains('suffix');
+        if (ispersonNameSuffixEnabled) {
+            dynamicSoql += ', Suffix ';
+        }
+
```
Comment (iskander-sfdc) [inline force-app/main/default/classes/UTIL_ACCT_Naming.cls:173]: Same here, not sure I understand this change in the scope of the test automation. Thanks

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls [hunk 3]
```diff
@@ -697,6 +703,11 @@ public class ACCT_IndividualAccounts_TDTM extends TDTM_Runnable {
             primaryConIdForAccount.put(acc.Primary_Contact__c, acc);
         }
 
+        // Clear cache to support real-time account model updates within the testing context.
+        if (Test.isRunningTest()) {
+            defaultRecTypeId = UTIL_CustomSettingsFacade.getSettings().Account_Processor__c;
+        }
+
```
Comment (iskander-sfdc) [inline force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:710]: Why do we need to alter the logic in the test context?

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 1 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:199]: These changes are required to automate [T-3904064](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeJIAU/view)/[T-3904063](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZeEIAU/view) as person name suffixes will currently result in an error.

Current behavior:

https://user-images.githubusercontent.com/16750742/152391837-219c74d6-9a46-4a46-8543-cbfe7f8acc0a.mp4


Proposed behavior:


https://user-images.githubusercontent.com/16750742/152391888-5dc12ddc-5984-4f83-8b5c-35c665b3e6b3.mp4

File: force-app/main/default/classes/UTIL_ACCT_Naming.cls (hunk 2 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/UTIL_ACCT_Naming.cls:173]: [Reference](https://github.com/SalesforceFoundation/EDA/pull/1950#discussion_r798784099)

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 3 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:710]: This change is relevant to [T-3904058](https://gus.lightning.force.com/lightning/r/ADM_Test_Scenario__c/a80B00000001ZduIAE/view). 

Given we are changing the default account model from household to administrative - the `defaultRecTypeId` was previously misreporting the value(because it is cached); This did not surface previously in our existing test suite because we haven't updated the default account model after setting it initially.

File: force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls [hunk 4]
```diff
@@ -445,7 +445,198 @@ private class ACCT_IndividualAccounts_TEST {
...
+            con.Email = 's@salesfor.com';
+            con.Preferred_Email__c = 'Correo electrónico laboral';
+
+            Test.startTest();
+            insert con;
+            Test.stopTest();
+
+            Account assertAccount = [SELECT Id, RecordType.Name, Name FROM Account LIMIT 1];
+            System.assertEquals('Household Account', assertAccount.RecordType.Name);
+            System.assertEquals('Domicilio de ' + con.LastName, assertAccount.Name);
```
Comment (salvadorsfdc) [inline force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls:478]: Hi @ZeeshanSyedH , I'm just getting started reviewing some of the changes, but I noticed this assertion passes, however I'm still wondering how this worked. I repeated the steps in my 2gp org with a user with Spanish language. The account name should be con.Lastname + 'Cuenta'.
This is because the Household_Other_Name_Setting__c is set to Label.lastNameHH (<value>{!LastName} Household</value>). Please take a look at the screenshot:
image

**Review by salvadorsfdc — COMMENTED** [review]
Some initial notes about some testing and findings

File: force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls (hunk 4 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls:478]: Within the test context here, since we consume the label as the running user(language: spanish) - the Household_Other_Name_Setting__c field was set too "Domicilio de {!LastName}" ([reference](https://github.com/SalesforceFoundation/EDA/blob/main/force-app/main/default/translations/es.translation-meta.xml#L252))

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 1 above)
Comment (salvadorsfdc) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:199]: Interesting, good point Iskander and Thanks @ZeeshanSyedH  for documenting it. 
I wonder if there is a known bug reported already about Suffixes not working correctly. I believe that in terms of the scope of this automation we possibly should not fix any bugs as well? I wonder if this is something we should bring up in our next stand up and let Jason and other devs decide.

File: force-app/main/default/classes/UTIL_ACCT_Naming.cls (hunk 2 above)
Comment (salvadorsfdc) [reply force-app/main/default/classes/UTIL_ACCT_Naming.cls:173]: as previously noted, I think we should ask our PM and other devs/leads about this, just to get more context about it and assess potential risks.

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 1 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:199]: I agree with Sal. I tested this change and it fixes the suffix issue indicated on the video here. I think we should discuss this with Jason

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 3 above)
Comment (iskander-sfdc) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:710]: I see, thanks for the explanation. Maybe instead of altering the logic here we could add `@TestVisible` annotation to `defaultRecTypeId` and change it from the test class? Just a suggestion
TDTM static variables is a problem, I saw similar problematic behavior in a few other places too

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 1 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls:199]: Agreed, lets take it up with product. For what its worth, the possibility of using suffix is mentioned within the product documentation currently:
Screen Shot 2022-02-03 at 7 23 29 PM

File: force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls (hunk 3 above)
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TDTM.cls]: Sure thing, that should also work ([reference commit](https://github.com/SalesforceFoundation/EDA/pull/1950/commits/905f6f23fa3b337318fa81e9ecf1bfd881e146a8))

File: force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls [hunk 5]
```diff
@@ -2696,4 +3120,25 @@ private class ACCT_IndividualAccounts_TEST {
 
         System.assertEquals('test 0 Administrative Account', accAfterUpdate.Name);
     }
-}
\ No newline at end of file
+
+    /*****************************************************************************************************************************
+     * @description Helper method to reurn a user in Spanish Language
+     *****************************************************************************************************************************/
+    private static User returnUserInSpanish() {
```
[the remark this answers is not in the document]
Comment (ZeeshanSyedH) [reply force-app/main/default/classes/ACCT_IndividualAccounts_TEST.cls]: Good idea ([reference commit](https://github.com/SalesforceFoundation/EDA/pull/1950/commits/efa3723425eec50b368892812a61421206a54a2a))

**Review by salvadorsfdc — APPROVED** [review]
looks good

**Review by iskander-sfdc — APPROVED** [review]
